utils/iscsi-pr.c

```diff
@@ -0,0 +1,476 @@
+/* 
+   Copyright (C) 2022 by zhenwei pi<pizhenwei@bytedance.com>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#ifdef HAVE_POLL_H
+#include <poll.h>
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <getopt.h>
+#include <inttypes.h>
+#include "iscsi.h"
+#include "scsi-lowlevel.h"
+
+static const char *initiator = "iqn.2022.com.github:pizhenwei:libiscsi:iscsi-pr";
+
+static void pr_in_read_keys(struct scsi_persistent_reserve_in_read_keys *rk)
+{
+	int i;
+
+	printf("  PR generation=0x%x, %d registered reservation keys follow:\n",
+			rk->prgeneration, rk->num_keys);
+	for (i = 0; i < rk->num_keys; i++) {
+		printf("    0x%" PRIx64 "\n", rk->keys[i]);
+	}
+}
+
+static void pr_in_read_reservation(struct scsi_persistent_reserve_in_read_reservation *rr)
+{
+	printf("  PR generation=0x%x, Reservation follows:\n", rr->prgeneration);
+	if (!rr->reserved) {
+		return;
+	}
+
+	printf("    Key=0x%" PRIx64 "\n", rr->reservation_key);
+	if (!rr->pr_scope) {
+		printf("    scope: LU_SCOPE,  type: %s\n", scsi_pr_type_str(rr->pr_type));
+	} else {
+		printf("    scope: %d,  type: %s\n",
+				rr->pr_scope, scsi_pr_type_str(rr->pr_type));
+	}
+}
+
+static void pr_in_report_capabilities(struct scsi_persistent_reserve_in_report_capabilities *rc)
+{
+	uint8_t  i;
+	uint16_t __mask = rc->persistent_reservation_type_mask;
+	uint16_t mask = ((__mask & 0xff) << 8) | (__mask >> 8);
```

Comment:
Please use scsi_get_uint*() for conversion from big endian to little endian.

Reply:
@bvanassche Hi, could you give me any hint about above questions? Or I just format code without tabs and force push a new version?